libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestClient.java

```diff
@@ -597,38 +601,87 @@ public HttpResponse performRequest(int method, URI url, HttpEntity httpEntity,
 			case Request.Method.PUT:
 				exec = httpAccessor.doPut(headers, url, httpEntity); break;
 			}
-			final HttpResponse response = exec.response;
-			int statusCode = response.getStatusLine().getStatusCode();
-
-			// 401 bad access token.
-			if (retryInvalidToken && statusCode == HttpStatus.SC_UNAUTHORIZED) {
-				final HttpEntity entity = response.getEntity();
-				if (entity != null && entity.isStreaming()) {
-					final InputStream instream = entity.getContent();
-					if (instream != null) {
-						instream.close();
-					}
-				}
-				refreshAccessToken();
-				return performRequest(method, url, httpEntity, additionalHttpHeaders, false);
-			}
 
-			// Done.
-			return response;
+			return exec.response;
 		}
 
 		/**
 		 * Swaps the existing access token for a new one.
 		 */
-		private void refreshAccessToken() {
-			if (authTokenProvider != null) {
+		private void refreshAccessToken(HttpResponse response) throws IOException {
+            // If we haven't retried already and we have an accessTokenProvider
+            // Then let's try to get a new authToken
+            if (authTokenProvider != null) {
+                // remember to consume this response so the connection can get re-used
+                HttpEntity entity = response.getEntity();
```

Comment:
Where's this coming from? I don't think we need ```consume()``` anymore, since we don't use the ```Apache HTTP client``` anymore.